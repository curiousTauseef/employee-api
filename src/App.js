import React, { Component } from 'react'
import styled from 'styled-components'
import logo from './logo.svg'
import './App.css'

const Wrapper = styled.div`
    text-align: center;
`

class App extends Component {
    render() {
        return (
            <Wrapper>
                <header className="App-header">
                    <img src={logo} className="App-logo" alt="logo" />
                    <h1>
                        Edit <code>src/App.js</code> and save to reload.
                    </h1>
                    <a
                        className="App-link"
                        href="https://reactjs.org"
                        target="_blank"
                        rel="noopener noreferrer"
                    >
                        Learn React
                    </a>
                </header>
            </Wrapper>
        )
    }
}

export default App
