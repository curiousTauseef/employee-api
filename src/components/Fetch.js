import { Component } from 'react'

export default class Fetch extends Component {
    state = {
        error: null,
        data: [],
        loading: true,
    }

    componentDidMount() {
        this.fetchEmployees()
            .catch(error => {
                console.log(error)
                return { error: error.message }
            })
            .then(data => {
                if (data.error) {
                    this.setState({ error: data.error, loading: false })
                } else {
                    this.setState({ data, loading: false })
                }
            })
    }

    fetchEmployees = async () => {
        const url = 'http://localhost:8000/api/employees'
        const response = await (await fetch(url)).json()
        return response
    }

    render() {
        const { children } = this.props
        return children({ ...this.state })
    }
}
