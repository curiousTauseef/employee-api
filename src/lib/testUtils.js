const mockData = [
    {
        uuid: '123',
        company: 'Streich PLC',
        bio: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        name: 'Benjamin Shields',
        title: 'Real Estate Sales Agent',
        avatar: 'http://httpstat.us/503',
    },
    {
        uuid: '456',
        company: 'Landgraab Industries',
        bio: '<p>Etiam ut mi sed neque elementum rhoncus.</p>',
        name: 'Gregory Landgraab',
        title: 'CEO',
        avatar: 0,
    },
    {
        uuid: '789',
        company: 'Streich PLC',
        bio: '<pre>Ut eget dui eleifend, pretium dui at, posuere dui.',
        name: 'Bella Goth',
        title: 'Test Subject',
        avatar: 'http://placekitten.com/50/50',
    },
]

module.exports = {
    mockData,
}
