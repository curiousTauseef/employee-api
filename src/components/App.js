import React from 'react'
import styled, { ThemeProvider, keyframes } from 'styled-components'
import logo from '../img/logo.svg'
import Fetch from './Fetch'
import Employee from './Employee'
import Pagination from './Pagination'
import theme from '../theme'
import GlobalStyle from './GlobalStyle'

const Wrapper = styled.div`
    text-align: center;
`

const Header = styled.header`
    margin: 40px 0;
`

const List = styled.ul`
    list-style: none;
    padding: 0;
    margin: 0 20px 20px 20px;
    text-align: left;
    background-color: ${props => props.theme.white};
    box-shadow: ${props => props.theme.boxShadow};
`

const spinner = keyframes`
    from {
        transform: rotate(0deg);
    }
    to {
        transform: rotate(360deg);
    }
`

const Loading = styled.p`
    display: flex;
    align-items: center;
    justify-content: center;

    img {
        animation: ${spinner} infinite 2s linear;
        width: 50px;
        display: block;
    }
`

const Main = styled.main`
    max-width: 768px;
    margin: 0 auto;
`

const App = () => (
    <ThemeProvider theme={theme}>
        <Wrapper>
            <GlobalStyle />
            <Header>
                <h1>Employee List</h1>
            </Header>
            <Main>
                <Fetch>
                    {({ data, error, loading }) => {
                        if (loading)
                            return (
                                <Loading>
                                    <span>
                                        <img src={logo} alt="Spinner" />
                                    </span>
                                    <span>Loading...</span>
                                </Loading>
                            )
                        if (error) return <p>{error}</p>
                        return (
                            <div>
                                <p>Showing {data.length} results</p>
                                <Pagination data={data} perPage={100}>
                                    {({ currentPage }) => (
                                        <div>
                                            SOME CHILD THING {currentPage}
                                        </div>
                                    )}
                                </Pagination>
                                <List>
                                    {data.map(employee => (
                                        <Employee
                                            key={employee.uuid}
                                            employee={employee}
                                        />
                                    ))}
                                </List>
                            </div>
                        )
                    }}
                </Fetch>
            </Main>
        </Wrapper>
    </ThemeProvider>
)

export default App

// <Pagination data={data}>
// {({
//     perPage,
//     pageCount,
//     currentPage,
//     next,
// }) => (
//     <div>
//         <p>perPage: {perPage}</p>
//         <p>pageCount: {pageCount}</p>
//         <p>currentPage: {currentPage}</p>
//         <button onClick={next}>NEXT</button>
//     </div>
// )}
// </Pagination>
