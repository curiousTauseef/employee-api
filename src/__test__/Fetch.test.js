import React from 'react'
import { mount } from 'enzyme'
import wait from 'waait'
import App from '../components/App'
import Fetch from '../components/Fetch'
import { mockData } from '../lib/testUtils'

describe('<Fetch />', () => {
    it('Renders employee data and matches snapshot', async () => {
        global.fetch = jest.fn().mockResolvedValue({
            json: () => mockData,
        })
        const wrapper = mount(<App />)
        await wait()
        wrapper.update()
        const fetchComponent = wrapper.find(Fetch)
        expect(global.fetch).toHaveBeenCalled()

        expect(fetchComponent).toMatchSnapshot()
        global.fetch.mockReset()
    })
})
