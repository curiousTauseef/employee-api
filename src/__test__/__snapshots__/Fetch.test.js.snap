// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fetch /> Renders an error message when fetch fails and matches snapshot 1`] = `
<Fetch>
  <p>
    There has been an error...
  </p>
</Fetch>
`;

exports[`<Fetch /> Renders employee data when successfully fetching and matches snapshot 1`] = `
<Fetch>
  <div>
    <p>
      Showing 
      3
       results
    </p>
    <Pagination
      data={
        Array [
          Object {
            "avatar": "http://httpstat.us/503",
            "bio": "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
            "company": "Streich PLC",
            "name": "Benjamin Shields",
            "title": "Real Estate Sales Agent",
            "uuid": "123",
          },
          Object {
            "avatar": 0,
            "bio": "<p>Etiam ut mi sed neque elementum rhoncus.</p>",
            "company": "Landgraab Industries",
            "name": "Gregory Landgraab",
            "title": "CEO",
            "uuid": "456",
          },
          Object {
            "avatar": "http://placekitten.com/50/50",
            "bio": "<pre>Ut eget dui eleifend, pretium dui at, posuere dui.",
            "company": "Streich PLC",
            "name": "Bella Goth",
            "title": "Test Subject",
            "uuid": "789",
          },
        ]
      }
      perPage={100}
    >
      <div>
        <button
          className="prev"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          PREV
        </button>
        <span>
          Showing page 
          1
           of 
          1
        </span>
        <button
          className="next"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          NEXT
        </button>
        <div>
          SOME CHILD THING 
          1
        </div>
      </div>
    </Pagination>
    <styled.ul>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bZQynM",
              "isStatic": false,
              "lastClassName": "ctKNap",
              "rules": Array [
                "
    list-style: none;
    padding: 0;
    margin: 0 20px 20px 20px;
    text-align: left;
    background-color: ",
                [Function],
                ";
    box-shadow: ",
                [Function],
                ";
",
              ],
            },
            "displayName": "styled.ul",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bZQynM",
            "target": "ul",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <ul
          className="sc-bZQynM ctKNap"
        >
          <Employee
            employee={
              Object {
                "avatar": "http://httpstat.us/503",
                "bio": "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
                "company": "Streich PLC",
                "name": "Benjamin Shields",
                "title": "Real Estate Sales Agent",
                "uuid": "123",
              }
            }
            key="123"
          >
            <styled.li>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bdVaJa",
                      "isStatic": true,
                      "lastClassName": "jOsBxC",
                      "rules": Array [
                        "
    display: flex;
    padding: 20px 0 0 20px;
",
                      ],
                    },
                    "displayName": "styled.li",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bdVaJa",
                    "target": "li",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <li
                  className="sc-bdVaJa jOsBxC"
                >
                  <styled.div>
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bwzfXH",
                            "isStatic": true,
                            "lastClassName": "eIhTQf",
                            "rules": Array [
                              "
    border-radius: 50%;
    flex-shrink: 0;
    overflow: hidden;
    width: 50px;
    height: 50px;
    object-fit: cover;
",
                            ],
                          },
                          "displayName": "styled.div",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-bwzfXH",
                          "target": "div",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                    >
                      <div
                        className="sc-bwzfXH eIhTQf"
                      >
                        <img
                          alt="Benjamin Shields"
                          onError={[Function]}
                          src="http://httpstat.us/503"
                          width="50px"
                        />
                      </div>
                    </StyledComponent>
                  </styled.div>
                  <styled.div>
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-htpNat",
                            "isStatic": false,
                            "lastClassName": "iGfYlM",
                            "rules": Array [
                              "
    margin-left: 20px;
    border-bottom: 1px solid ",
                              [Function],
                              ";
    flex-grow: 1;
    padding: 0 20px 20px 0;
",
                            ],
                          },
                          "displayName": "styled.div",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-htpNat",
                          "target": "div",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                    >
                      <div
                        className="sc-htpNat iGfYlM"
                      >
                        <h2>
                          Benjamin Shields
                        </h2>
                        <p>
                          Real Estate Sales Agent
                           at 
                          Streich PLC
                        </p>
                        <styled.p>
                          <StyledComponent
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-bxivhb",
                                  "isStatic": false,
                                  "lastClassName": "hJmXeG",
                                  "rules": Array [
                                    "
    font-size: 0.875rem;
    color: ",
                                    [Function],
                                    ";
",
                                  ],
                                },
                                "displayName": "styled.p",
                                "foldedComponentIds": Array [],
                                "render": [Function],
                                "styledComponentId": "sc-bxivhb",
                                "target": "p",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                          >
                            <p
                              className="sc-bxivhb hJmXeG"
                            >
                              Lorem ipsum dolor sit amet, consectetur adipiscing elit.
                            </p>
                          </StyledComponent>
                        </styled.p>
                      </div>
                    </StyledComponent>
                  </styled.div>
                </li>
              </StyledComponent>
            </styled.li>
          </Employee>
          <Employee
            employee={
              Object {
                "avatar": 0,
                "bio": "<p>Etiam ut mi sed neque elementum rhoncus.</p>",
                "company": "Landgraab Industries",
                "name": "Gregory Landgraab",
                "title": "CEO",
                "uuid": "456",
              }
            }
            key="456"
          >
            <styled.li>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bdVaJa",
                      "isStatic": true,
                      "lastClassName": "jOsBxC",
                      "rules": Array [
                        "
    display: flex;
    padding: 20px 0 0 20px;
",
                      ],
                    },
                    "displayName": "styled.li",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bdVaJa",
                    "target": "li",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <li
                  className="sc-bdVaJa jOsBxC"
                >
                  <styled.div>
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bwzfXH",
                            "isStatic": true,
                            "lastClassName": "eIhTQf",
                            "rules": Array [
                              "
    border-radius: 50%;
    flex-shrink: 0;
    overflow: hidden;
    width: 50px;
    height: 50px;
    object-fit: cover;
",
                            ],
                          },
                          "displayName": "styled.div",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-bwzfXH",
                          "target": "div",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                    >
                      <div
                        className="sc-bwzfXH eIhTQf"
                      >
                        <img
                          alt="Gregory Landgraab"
                          onError={[Function]}
                          src={0}
                          width="50px"
                        />
                      </div>
                    </StyledComponent>
                  </styled.div>
                  <styled.div>
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-htpNat",
                            "isStatic": false,
                            "lastClassName": "iGfYlM",
                            "rules": Array [
                              "
    margin-left: 20px;
    border-bottom: 1px solid ",
                              [Function],
                              ";
    flex-grow: 1;
    padding: 0 20px 20px 0;
",
                            ],
                          },
                          "displayName": "styled.div",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-htpNat",
                          "target": "div",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                    >
                      <div
                        className="sc-htpNat iGfYlM"
                      >
                        <h2>
                          Gregory Landgraab
                        </h2>
                        <p>
                          CEO
                           at 
                          Landgraab Industries
                        </p>
                        <styled.p>
                          <StyledComponent
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-bxivhb",
                                  "isStatic": false,
                                  "lastClassName": "hJmXeG",
                                  "rules": Array [
                                    "
    font-size: 0.875rem;
    color: ",
                                    [Function],
                                    ";
",
                                  ],
                                },
                                "displayName": "styled.p",
                                "foldedComponentIds": Array [],
                                "render": [Function],
                                "styledComponentId": "sc-bxivhb",
                                "target": "p",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                          >
                            <p
                              className="sc-bxivhb hJmXeG"
                            >
                              Etiam ut mi sed neque elementum rhoncus.
                            </p>
                          </StyledComponent>
                        </styled.p>
                      </div>
                    </StyledComponent>
                  </styled.div>
                </li>
              </StyledComponent>
            </styled.li>
          </Employee>
          <Employee
            employee={
              Object {
                "avatar": "http://placekitten.com/50/50",
                "bio": "<pre>Ut eget dui eleifend, pretium dui at, posuere dui.",
                "company": "Streich PLC",
                "name": "Bella Goth",
                "title": "Test Subject",
                "uuid": "789",
              }
            }
            key="789"
          >
            <styled.li>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bdVaJa",
                      "isStatic": true,
                      "lastClassName": "jOsBxC",
                      "rules": Array [
                        "
    display: flex;
    padding: 20px 0 0 20px;
",
                      ],
                    },
                    "displayName": "styled.li",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bdVaJa",
                    "target": "li",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <li
                  className="sc-bdVaJa jOsBxC"
                >
                  <styled.div>
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bwzfXH",
                            "isStatic": true,
                            "lastClassName": "eIhTQf",
                            "rules": Array [
                              "
    border-radius: 50%;
    flex-shrink: 0;
    overflow: hidden;
    width: 50px;
    height: 50px;
    object-fit: cover;
",
                            ],
                          },
                          "displayName": "styled.div",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-bwzfXH",
                          "target": "div",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                    >
                      <div
                        className="sc-bwzfXH eIhTQf"
                      >
                        <img
                          alt="Bella Goth"
                          onError={[Function]}
                          src="http://placekitten.com/50/50"
                          width="50px"
                        />
                      </div>
                    </StyledComponent>
                  </styled.div>
                  <styled.div>
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-htpNat",
                            "isStatic": false,
                            "lastClassName": "iGfYlM",
                            "rules": Array [
                              "
    margin-left: 20px;
    border-bottom: 1px solid ",
                              [Function],
                              ";
    flex-grow: 1;
    padding: 0 20px 20px 0;
",
                            ],
                          },
                          "displayName": "styled.div",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-htpNat",
                          "target": "div",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                    >
                      <div
                        className="sc-htpNat iGfYlM"
                      >
                        <h2>
                          Bella Goth
                        </h2>
                        <p>
                          Test Subject
                           at 
                          Streich PLC
                        </p>
                        <styled.p>
                          <StyledComponent
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-bxivhb",
                                  "isStatic": false,
                                  "lastClassName": "hJmXeG",
                                  "rules": Array [
                                    "
    font-size: 0.875rem;
    color: ",
                                    [Function],
                                    ";
",
                                  ],
                                },
                                "displayName": "styled.p",
                                "foldedComponentIds": Array [],
                                "render": [Function],
                                "styledComponentId": "sc-bxivhb",
                                "target": "p",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                          >
                            <p
                              className="sc-bxivhb hJmXeG"
                            >
                              Ut eget dui eleifend, pretium dui at, posuere dui.
                            </p>
                          </StyledComponent>
                        </styled.p>
                      </div>
                    </StyledComponent>
                  </styled.div>
                </li>
              </StyledComponent>
            </styled.li>
          </Employee>
        </ul>
      </StyledComponent>
    </styled.ul>
  </div>
</Fetch>
`;
